metadata-ingestion/src/datahub/ingestion/source/thrift.py

```diff
@@ -0,0 +1,141 @@
+    ) -> Generator[MetadataChangeProposalWrapper, None, None]:
+        for definition in ctx.definition():
+            yield from self.bind_MCPs_from_Definition(definition)
+
+    def bind_MCPs_from_Definition(
+        self, ctx: ThriftParser.DefinitionContext
+    ) -> Generator[MetadataChangeProposalWrapper, None, None]:
+        if ctx.struct_():
+            yield from self.bind_MCPs_from_struct_(ctx.struct_())
+        else:
+            raise NotImplementedError()
+
+    def bind_MCPs_from_struct_(
+        self, ctx: ThriftParser.Struct_Context
+    ) -> Generator[MetadataChangeProposalWrapper, None, None]:
+
+        name = ctx.IDENTIFIER().getText()
+        urn = f"urn:li:dataset:(urn:li:dataPlatform:thrift,{name},PROD)"
+        yield MetadataChangeProposalWrapper(
+            entityType="dataset",
+            changeType=ChangeType.UPSERT,
+            aspectName="schemaMetadata",
+            entityUrn=urn,
+            aspect=SchemaMetadata(
+                schemaName=name,
+                platform="urn:li:dataPlatform:thrift",
+                version=0,
+                hash="",
+                platformSchema=OtherSchema(""),
+                fields=[
+                    self.bind_SchemaField_from_field(field) for field in ctx.field()
+                ],
+            ),
+        )
+
+    def bind_SchemaField_from_field(
+        self, ctx: ThriftParser.FieldContext
+    ) -> SchemaField:
+        return SchemaField(
+            fieldPath=ctx.IDENTIFIER().getText(),
+            type=self.bind_SchemaFieldDataType_from_field_type(ctx.field_type()),
+            nativeDataType=self.bind_nativeDataType_from_field_type(ctx.field_type()),
+        )
+
+    def bind_nativeDataType_from_field_type(
+        self, ctx: ThriftParser.Field_typeContext
+    ) -> str:
+        if ctx.base_type():
+            return ctx.getText()
+        else:
+            raise NotImplementedError()
+
+    def bind_SchemaFieldDataType_from_field_type(
+        self, ctx: ThriftParser.Field_typeContext
+    ) -> SchemaFieldDataType:
+        if ctx.base_type():
+            return self.bind_SchemaFieldDataType_from_base_type(ctx.base_type())
+        else:
+            raise NotImplementedError()
+
+    def bind_SchemaFieldDataType_from_base_type(
+        self, ctx: ThriftParser.Base_typeContext
+    ) -> SchemaFieldDataType:
+        return self.bind_SchemaFieldDataType_from_Real_base_type(ctx.real_base_type())
+
+    def bind_SchemaFieldDataType_from_Real_base_type(
+        self, ctx: ThriftParser.Real_base_typeContext
+    ) -> SchemaFieldDataType:
+        if ctx.TYPE_I32() or ctx.TYPE_I64() or ctx.TYPE_DOUBLE():
+            return SchemaFieldDataType(NumberType())
+        elif ctx.TYPE_STRING:
+            return SchemaFieldDataType(StringType())
+        else:
+            raise NotImplementedError()
+
+
+@dataclass
+class ThriftSource(Source):
+    config: ThriftSourceConfig
+    report: SourceReport = field(default_factory=SourceReport)
```

Comment:
Are there any interesting metrics that we want to report, such as errors and processing time?

Reply:
This requirement will be resolved in following diffs.

Reply:
You can create a sub-class of SourceReport, called ThriftReport to extend it.

Reply:
You can look at [this example](https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/ingestion/source_report/sql/bigquery.py). There are more under [this](https://github.com/datahub-project/datahub/tree/master/metadata-ingestion/src/datahub/ingestion/source_report).

Reply:
I have created a sub-class ThriftReport and add variables: start time and end time. I found that failures{} has been in SourceReport, so I don't need to add errors{} to ThriftReport, yes?

Reply:
Correct.